lib/trino-collect/src/test/java/io/trino/collect/cache/TestSafeCaches.java

```diff
@@ -0,0 +1,93 @@
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.collect.ImmutableList;
+import org.testng.annotations.Test;
+
+import static io.trino.collect.cache.SafeCaches.buildNonEvictableCache;
+import static io.trino.collect.cache.SafeCaches.buildNonEvictableCacheWithWeakInvalidateAll;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+public class TestSafeCaches
+{
+    @Test
+    public void testNonEvictableCache()
+            throws Exception
+    {
+        NonEvictableCache<Object, Object> cache = buildNonEvictableCache(CacheBuilder.newBuilder());
+        assertThatKeyInvalidationIsImpossible(cache);
+        assertThatClearIsImpossible(cache);
+        assertThatLoadingIsPossible(cache);
+    }
+
+    @Test
+    public void testNonEvictableCacheWithWeakInvalidateAll()
+            throws Exception
+    {
+        NonKeyEvictableCache<Object, Object> cache = buildNonEvictableCacheWithWeakInvalidateAll(CacheBuilder.newBuilder());
+        assertThatKeyInvalidationIsImpossible(cache);
+        assertThatClearIsPossible(cache);
+        assertThatLoadingIsPossible(cache);
+    }
+
+    @Test
+    public void testNonEvictableLoadingCache()
+            throws Exception
+    {
+        NonEvictableLoadingCache<Object, Object> cache = buildNonEvictableCache(CacheBuilder.newBuilder(), CacheLoader.from(key -> key));
+        assertThatKeyInvalidationIsImpossible(cache);
+        assertThatClearIsImpossible(cache);
+        assertThatLoadingIsPossible(cache);
+    }
+
+    @Test
+    public void testNonEvictableLoadingCacheWithWeakInvalidateAll()
+            throws Exception
+    {
+        NonKeyEvictableLoadingCache<Object, Object> cache = buildNonEvictableCacheWithWeakInvalidateAll(CacheBuilder.newBuilder(), CacheLoader.from(key -> key));
+        assertThatKeyInvalidationIsImpossible(cache);
+        assertThatClearIsPossible(cache);
+        assertThatLoadingIsPossible(cache);
+    }
+
+    private static void assertThatLoadingIsPossible(Cache<Object, Object> cache)
+            throws Exception
+    {
+        Object object = new Object();
+        cache.get(object, () -> object);
+    }
+
+    private static void assertThatKeyInvalidationIsImpossible(Cache<Object, Object> cache)
+    {
+        Object object = new Object();
+        assertThatThrownBy(() -> cache.invalidate(object)).isInstanceOf(UnsupportedOperationException.class);
+        assertThatThrownBy(() -> cache.invalidateAll(ImmutableList.of())).isInstanceOf(UnsupportedOperationException.class);
+    }
+
+    private static void assertThatClearIsImpossible(Cache<Object, Object> cache)
+    {
+        assertThatThrownBy(cache::invalidateAll).isInstanceOf(UnsupportedOperationException.class);
+        // TODO test asMap().clear()
+    }
+
+    private static void assertThatClearIsPossible(Cache<Object, Object> cache)
+    {
+        cache.invalidateAll();
+        cache.asMap().clear();
```

Comment:
add assertion that cache gets emptied by each of these

Reply:
> So testing if methods are properly implemented should be on a different level

we don't have tests for SafeCaches's wrappers, which was a mistake. You're adding them now.

> where a delegate cache implementation is.

We could want a "smoke test" to verify things generally work. I.e. what if SafeCaches wrapper missed a call to delegate?

Other than that, I agree there is no point in fine grained tests, we don't want to test underlying implementation here, nor concurrency, etc.

Reply:
> but the challenge is that it is not clear how to put/load data into the cache.

maybe we can have something like


```java
private static void verifyClearIsPossible(Cache<Object, Object> cache)
        throws Exception
{
    Object key = new Object();
    Object firstValue = new Object();
    cache.get(key, () -> firstValue);
    assertSame(cache.getIfPresent(key), firstValue);

    cache.invalidateAll();
    assertNull(cache.getIfPresent(key));

    Object secondValue = new Object();
    cache.get(key, () -> secondValue);
    assertSame(cache.getIfPresent(key), secondValue);
    cache.asMap().clear();
    assertNull(cache.getIfPresent(key));
}
```